File: 2020/day7.cpp
#include <iostream>
#include <regex>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "adventofcode.h"

typedef struct
{
	std::string name;
	int number;
} bag_rule_t;

class Day7
{
	std::unordered_set<std::string> compatible;
	std::unordered_map<std::string, std::vector<bag_rule_t>> bag_rules;

public:
	int run(std::istream &file, bool part1)
	{
		std::string line;
		std::regex parent_reg(R"(([\w ]+?) bags contain)"); // captures the name of the parent bag
		std::regex child_reg(R"((\d+) ([\w ]+?) bag)");		// captures the name and count of the child bags

		while (std::getline(file, line))
		{
			std::smatch matches;
			if (!std::regex_search(line, matches, parent_reg))
				continue;

			auto parent = matches[1].str();
			std::string suffix = matches.suffix().str();
			while (std::regex_search(suffix, matches, child_reg))
			{
				int count = std::stoi(matches[1].str());
				auto child = matches[2].str();
				if (part1)
					bag_rules[child].push_back({parent, count});
				else
					bag_rules[parent].push_back({child, count});
				suffix = matches.suffix().str();
			}
		}

		auto rules = bag_rules["shiny gold"];

		if (part1)
		{
			find_all_compatible(rules);
			return compatible.size();
		}
		else
		{
			return find_all_required(rules) - 1; // not counting the bag itself
		}
	}

private:
	void find_all_compatible(std::vector<bag_rule_t> &rules)
	{
		for (auto &rule : rules)
		{
			compatible.insert(rule.name);
			find_all_compatible(bag_rules[rule.name]);
		}
	}

	int find_all_required(std::vector<bag_rule_t> &rules)
	{
		int required = 1; // counting the bag itself
		// now find all bags required within this bag
		for (auto &rule : rules)
		{
			required += rule.number * find_all_required(bag_rules[rule.name]);
		}
		return required;
	}
};

int day7_1(std::istream &file)
{
	Day7 solution;
	std::cout << solution.run(file, true) << std::endl;
	return 0;
}

int day7_2(std::istream &file)
{
	Day7 solution;
	std::cout << solution.run(file, false) << std::endl;
	return 0;
}

File: 2020/rust/src/day7.rs
use crate::solver::Solver;
use crate::util::push_key;
use regex::Regex;
use std::collections::{HashMap, HashSet, VecDeque};

#[derive(Clone)]
struct BagRule {
    name: String,
    number: i32,
}

#[derive(Default)]
pub struct Day7 {
    compatible: HashSet<String>,
    child_to_parent: HashMap<String, Vec<BagRule>>,
    parent_to_child: HashMap<String, Vec<BagRule>>,
}

impl Day7 {
    // cannot use recursion due to needing a mutable reference to self
    fn find_all_compatible(&mut self, rules: &Vec<BagRule>) {
        let mut work = VecDeque::new();
        work.push_back(rules);
        while let Some(rules) = work.pop_front() {
            for ref rule in rules {
                self.compatible.insert(rule.name.clone());
                if let Some(parent_rules) = self.child_to_parent.get(&rule.name) {
                    work.push_back(parent_rules);
                }
            }
        }
    }

    fn find_all_required(&self, rules: &Vec<BagRule>) -> i32 {
        let mut required = 1;
        for rule in rules {
            required += rule.number
                * match self.parent_to_child.get(&rule.name) {
                    Some(rules) => self.find_all_required(rules),
                    None => 1,
                };
        }
        required
    }

    fn solve_part1(&mut self) -> i64 {
        let rules = self
            .child_to_parent
            .get("shiny gold")
            .unwrap_or(&Vec::new())
            .clone();
        self.find_all_compatible(&rules);
        return self.compatible.len() as i64;
    }

    fn solve_part2(&self) -> i64 {
        let rules = self
            .parent_to_child
            .get("shiny gold")
            .unwrap_or(&Vec::new())
            .clone();
        return self.find_all_required(&rules) as i64 - 1;
    }
}

impl Solver for Day7 {
    fn new(lines: impl Iterator<Item = String>) -> Self {
        let mut day7 = Self::default();
        let parent_reg = Regex::new(r"([\w ]+?) bags contain").unwrap();
        let child_reg = Regex::new(r"(\d+) ([\w ]+?) bag").unwrap();

        for ref line in lines {
            let matches = match parent_reg.captures(line) {
                Some(matches) => matches,
                None => continue,
            };

            let parent = matches.get(1).unwrap().as_str().to_owned();
            let mut suffix = &line[matches.get(0).unwrap().end()..];
            while let Some(matches) = child_reg.captures(suffix) {
                let count = matches
                    .get(1)
                    .unwrap()
                    .as_str()
                    .parse::<i32>()
                    .unwrap_or_default();
                let child = matches.get(2).unwrap().as_str().to_owned();
                push_key(
                    &mut day7.child_to_parent,
                    child.clone(),
                    BagRule {
                        name: parent.clone(),
                        number: count,
                    },
                );
                push_key(
                    &mut day7.parent_to_child,
                    parent.clone(),
                    BagRule {
                        name: child.clone(),
                        number: count,
                    },
                );
                suffix = &suffix[matches.get(0).unwrap().end()..];
            }
        }

        day7
    }

    fn solve(&mut self) -> (i64, i64) {
        (self.solve_part1(), self.solve_part2())
    }
}
